// @iplab/ngx-file-upload/esm5/lib/services/file-upload.service.js
import { __decorate, __metadata, __read } from "tslib";
import { Injectable, Renderer2 } from '@angular/core';
import { FileUploadTypes } from './../helpers/file-types.class';
import { IsNullOrEmpty } from './../helpers/helpers.class';
import * as ɵngcc0 from '@angular/core';
var FileUploadService = /** @class */ (function () {
    function FileUploadService(renderer) {
        this.renderer = renderer;
        this.extensions = ['B', 'KB', 'MB', 'GB'];
        this.sizeRegex = new RegExp("^(\\d+)(?:\\s{0,1})(" + this.extensions.join('|') + ")?$", 'i');
    }
    FileUploadService.prototype.isFileDragDropAvailable = function () {
        var div = this.renderer.createElement('div');
        return (('draggable' in div) || ('ondragstart' in div && 'ondrop' in div));
    };
    FileUploadService.prototype.parseSize = function (value) {
        if (IsNullOrEmpty(value)) {
            return 0;
        }
        if (typeof value === 'number') {
            return value;
        }
        var _a = __read(value.match(this.sizeRegex) || [null, '0', 'B'], 3), size = _a[1], extension = _a[2];
        var i = IsNullOrEmpty(extension) ? 0 : this.extensions.indexOf(extension.toUpperCase());
        return parseInt(size, 10) * Math.pow(1024, i < 0 ? 0 : i);
    };
    FileUploadService.prototype.formatSize = function (size) {
        return this.calculateSize(size);
    };
    FileUploadService.prototype.calculateSize = function (size, extensionIndex) {
        if (extensionIndex === void 0) { extensionIndex = 0; }
        if (isNaN(size)) {
            size = 0;
        }
        if (size < 1024) {
            return Math.round(size * 100) / 100 + " " + this.extensions[extensionIndex];
        }
        return this.calculateSize(size / 1024, extensionIndex + 1);
    };
    FileUploadService.prototype.getFileType = function (file) {
        return Object.keys(FileUploadTypes).find(function (key) { return FileUploadTypes[key] === file.type; });
    };
    FileUploadService.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    FileUploadService = __decorate([ __metadata("design:paramtypes", [Renderer2])
    ], FileUploadService);
FileUploadService.ɵfac = function FileUploadService_Factory(t) { return new (t || FileUploadService)(ɵngcc0.ɵɵinject(ɵngcc0.Renderer2)); };
FileUploadService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: FileUploadService, factory: function (t) { return FileUploadService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FileUploadService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc0.Renderer2 }]; }, null); })();
    return FileUploadService;
}());
export { FileUploadService };
