// angular-highcharts-chart/esm5/lib/angular-highcharts-chart.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { AngularHighchartsChartComponent } from './angular-highcharts-chart.component';
import * as ɵngcc0 from '@angular/core';
var AngularHighchartsChartModule = /** @class */ (function () {
    function AngularHighchartsChartModule() {
    }
AngularHighchartsChartModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularHighchartsChartModule });
AngularHighchartsChartModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularHighchartsChartModule_Factory(t) { return new (t || AngularHighchartsChartModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularHighchartsChartModule, { declarations: function () { return [AngularHighchartsChartComponent]; }, exports: function () { return [AngularHighchartsChartComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularHighchartsChartModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [AngularHighchartsChartComponent],
                exports: [AngularHighchartsChartComponent]
            }]
    }], function () { return []; }, null); })();
    return AngularHighchartsChartModule;
}());
export { AngularHighchartsChartModule };
