// angular-highcharts-chart/esm5/lib/angular-highcharts-chart.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var AngularHighchartsChartComponent = /** @class */ (function () {
    function AngularHighchartsChartComponent(el) {
        this.el = el;
        this.updateValue = false;
        this.updateChange = new EventEmitter(true);
        this.updateOrCreateChart = function () {
            if (this.chart && this.chart.update) {
                this.chart.update(this.optionsValue, true, this.oneToOne || false);
            }
            else {
                this.chart = this.Highcharts[this.constructorType || 'chart'](this.el.nativeElement, this.optionsValue, this.callbackFunction || null);
                this.optionsValue.series = this.chart.userOptions.series;
            }
        };
    }
    Object.defineProperty(AngularHighchartsChartComponent.prototype, "options", {
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this.optionsValue = val;
            this.updateOrCreateChart();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AngularHighchartsChartComponent.prototype, "update", {
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val) {
                this.updateOrCreateChart();
                this.updateChange.emit(false); // clear the flag after update
            }
        },
        enumerable: true,
        configurable: true
    });
    /** @nocollapse */
    AngularHighchartsChartComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    AngularHighchartsChartComponent.propDecorators = {
        Highcharts: [{ type: Input }],
        constructorType: [{ type: Input }],
        callbackFunction: [{ type: Input }],
        options: [{ type: Input }],
        updateChange: [{ type: Output }],
        update: [{ type: Input }],
        oneToOne: [{ type: Input }]
    };
AngularHighchartsChartComponent.ɵfac = function AngularHighchartsChartComponent_Factory(t) { return new (t || AngularHighchartsChartComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AngularHighchartsChartComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AngularHighchartsChartComponent, selectors: [["lib-highcharts-angular"]], inputs: { options: "options", update: "update", Highcharts: "Highcharts", constructorType: "constructorType", callbackFunction: "callbackFunction", oneToOne: "oneToOne" }, outputs: { updateChange: "updateChange" }, decls: 0, vars: 0, template: function AngularHighchartsChartComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularHighchartsChartComponent, [{
        type: Component,
        args: [{
                selector: 'lib-highcharts-angular',
                template: ''
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { updateChange: [{
            type: Output
        }], options: [{
            type: Input
        }], update: [{
            type: Input
        }], Highcharts: [{
            type: Input
        }], constructorType: [{
            type: Input
        }], callbackFunction: [{
            type: Input
        }], oneToOne: [{
            type: Input
        }] }); })();
    return AngularHighchartsChartComponent;
}());
export { AngularHighchartsChartComponent };
function AngularHighchartsChartComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    AngularHighchartsChartComponent.prototype.chart;
    /** @type {?} */
    AngularHighchartsChartComponent.prototype.Highcharts;
    /** @type {?} */
    AngularHighchartsChartComponent.prototype.constructorType;
    /** @type {?} */
    AngularHighchartsChartComponent.prototype.callbackFunction;
    /** @type {?} */
    AngularHighchartsChartComponent.prototype.optionsValue;
    /** @type {?} */
    AngularHighchartsChartComponent.prototype.updateValue;
    /** @type {?} */
    AngularHighchartsChartComponent.prototype.updateChange;
    /** @type {?} */
    AngularHighchartsChartComponent.prototype.oneToOne;
    /** @type {?} */
    AngularHighchartsChartComponent.prototype.updateOrCreateChart;
    /** @type {?} */
    AngularHighchartsChartComponent.prototype.el;
}
