// angular-dual-listbox/esm2015/lib/angular-dual-listbox.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { DualListComponent } from './dual-list.component';
import * as ɵngcc0 from '@angular/core';
export class AngularDualListBoxModule {
}
AngularDualListBoxModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularDualListBoxModule });
AngularDualListBoxModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularDualListBoxModule_Factory(t) { return new (t || AngularDualListBoxModule)(); }, imports: [[
            CommonModule,
            FormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularDualListBoxModule, { declarations: function () { return [DualListComponent]; }, imports: function () { return [CommonModule,
        FormsModule]; }, exports: function () { return [DualListComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularDualListBoxModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule
                ],
                declarations: [DualListComponent],
                exports: [DualListComponent]
            }]
    }], null, null); })();
