// angular2-tinymce/esm2015/lib/angular2-tinymce-lib.default.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
export class TinymceDefaultOptions {
    constructor() {
        this.plugins = [
            'link',
            'paste',
            'table',
            'advlist',
            'autoresize',
            'lists',
            'code'
        ];
        this.skin_url = '/assets/tinymce/skins/lightgray';
        this.baseURL = '/assets/tinymce';
        this.auto_focus = true;
    }
}
TinymceDefaultOptions.ɵfac = function TinymceDefaultOptions_Factory(t) { return new (t || TinymceDefaultOptions)(); };
TinymceDefaultOptions.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: TinymceDefaultOptions, factory: TinymceDefaultOptions.ɵfac });

if (false) {
    /** @type {?} */
    TinymceDefaultOptions.prototype.plugins;
    /** @type {?} */
    TinymceDefaultOptions.prototype.skin_url;
    /** @type {?} */
    TinymceDefaultOptions.prototype.baseURL;
    /** @type {?} */
    TinymceDefaultOptions.prototype.auto_focus;
}
