// @iplab/ngx-file-upload/esm5/lib/directives/attribute.directive.js
import { __decorate, __metadata, __param } from "tslib";
import { Directive, Input, Optional, Host, Self } from '@angular/core';
import { FileUploadComponent } from '../components/multiple-file-upload/file-upload.component';
import { SimpleFileUploadComponent } from '../components/simple-file-upload/simple-file-upload.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../components/multiple-file-upload/file-upload.component';
import * as ɵngcc2 from '../components/simple-file-upload/simple-file-upload.component';
var FilesAcceptDirective = /** @class */ (function () {
    function FilesAcceptDirective(fileUpload, simpleFileUpload) {
        this.fileUpload = null;
        this.fileUpload = fileUpload || simpleFileUpload;
    }
    FilesAcceptDirective.prototype.ngAfterViewInit = function () {
        this.setAccept(this.accept);
    };
    FilesAcceptDirective.prototype.ngOnChanges = function (changes) {
        if ('accept' in changes && changes['accept'].currentValue !== changes['accept'].previousValue) {
            this.setAccept(this.accept);
        }
    };
    FilesAcceptDirective.prototype.setAccept = function (accept) {
        if (this.fileUpload && this.fileUpload.control) {
            this.fileUpload.control.acceptFiles(accept);
        }
    };
    FilesAcceptDirective.ctorParameters = function () { return [
        { type: FileUploadComponent, decorators: [{ type: Optional }, { type: Host }, { type: Self }] },
        { type: SimpleFileUploadComponent, decorators: [{ type: Optional }, { type: Host }, { type: Self }] }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], FilesAcceptDirective.prototype, "accept", void 0);
    FilesAcceptDirective = __decorate([ __param(0, Optional()), __param(0, Host()), __param(0, Self()),
        __param(1, Optional()), __param(1, Host()), __param(1, Self()),
        __metadata("design:paramtypes", [FileUploadComponent,
            SimpleFileUploadComponent])
    ], FilesAcceptDirective);
FilesAcceptDirective.ɵfac = function FilesAcceptDirective_Factory(t) { return new (t || FilesAcceptDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FileUploadComponent, 11), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.SimpleFileUploadComponent, 11)); };
FilesAcceptDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: FilesAcceptDirective, selectors: [["file-upload", "accept", ""]], hostVars: 1, hostBindings: function FilesAcceptDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵattribute("accept", ctx.accept ? ctx.accept : null);
    } }, inputs: { accept: "accept" }, features: [ɵngcc0.ɵɵNgOnChangesFeature] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FilesAcceptDirective, [{
        type: Directive,
        args: [{
                selector: 'file-upload[accept]',
                host: { '[attr.accept]': 'accept ? accept : null' }
            }]
    }], function () { return [{ type: ɵngcc1.FileUploadComponent, decorators: [{
                type: Optional
            }, {
                type: Host
            }, {
                type: Self
            }] }, { type: ɵngcc2.SimpleFileUploadComponent, decorators: [{
                type: Optional
            }, {
                type: Host
            }, {
                type: Self
            }] }]; }, { accept: [{
            type: Input
        }] }); })();
    return FilesAcceptDirective;
}());
export { FilesAcceptDirective };
