// angular-pell/bundles/angular-pell.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common')) :
	typeof define === 'function' && define.amd ? define(['@angular/core','@angular/common','exports', '@angular/core', '@angular/common'], factory) :
	(factory(global.ng.core,global.ng.common,(global['angular-pell'] = {}),global.ng.core,global.ng.common));
}(this, (function (ɵngcc0,ɵngcc1,exports,core,common) { 
function PellComponent_button_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 3);
    ɵngcc0.ɵɵlistener("click", function PellComponent_button_3_Template_button_click_0_listener() { var action_r2 = ctx.$implicit; return action_r2.result(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var action_r2 = ctx.$implicit;
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassMap(ctx_r1.actionButtonClass);
    ɵngcc0.ɵɵproperty("title", action_r2.title)("innerHTML", action_r2.icon, ɵngcc0.ɵɵsanitizeHtml);
} }
'use strict';

var actions = {
    bold: {
        icon: '<b>B</b>',
        title: 'Bold',
        result: function () { return exec('bold'); }
    },
    italic: {
        icon: '<i>I</i>',
        title: 'Italic',
        result: function () { return exec('italic'); }
    },
    underline: {
        icon: '<u>U</u>',
        title: 'Underline',
        result: function () { return exec('underline'); }
    },
    strikethrough: {
        icon: '<strike>S</strike>',
        title: 'Strike-through',
        result: function () { return exec('strikeThrough'); }
    },
    heading1: {
        icon: '<b>H<sub>1</sub></b>',
        title: 'Heading 1',
        result: function () { return exec('formatBlock', '<H1>'); }
    },
    heading2: {
        icon: '<b>H<sub>2</sub></b>',
        title: 'Heading 2',
        result: function () { return exec('formatBlock', '<H2>'); }
    },
    paragraph: {
        icon: '&#182;',
        title: 'Paragraph',
        result: function () { return exec('formatBlock', '<P>'); }
    },
    quote: {
        icon: '&#8220; &#8221;',
        title: 'Quote',
        result: function () { return exec('formatBlock', '<BLOCKQUOTE>'); }
    },
    olist: {
        icon: '&#35;',
        title: 'Ordered List',
        result: function () { return exec('insertOrderedList'); }
    },
    ulist: {
        icon: '&#8226;',
        title: 'Unordered List',
        result: function () { return exec('insertUnorderedList'); }
    },
    code: {
        icon: '&lt;/&gt;',
        title: 'Code',
        result: function () { return exec('formatBlock', '<PRE>'); }
    },
    line: {
        icon: '&#8213;',
        title: 'Horizontal Line',
        result: function () { return exec('insertHorizontalRule'); }
    },
    link: {
        icon: '&#128279;',
        title: 'Link',
        result: function () {
            var url = window.prompt('Enter the link URL');
            if (url)
                exec('createLink', url);
        }
    },
    image: {
        icon: '&#128247;',
        title: 'Image',
        result: function () {
            var url = window.prompt('Enter the image URL');
            if (url)
                exec('insertImage', url);
        }
    }
};
var exec = function (command, value) {
    if (value === void 0) { value = null; }
    document.execCommand(command, false, value);
};
var PellComponent = /** @class */ (function () {
    function PellComponent() {
        this.actions = Object.keys(actions).map(function (action) { return actions[action]; });
        this.actionBarClass = 'pell-actionbar';
        this.actionButtonClass = 'pell-button';
        this.contentClass = 'pell-content';
        this.onChange = new core.EventEmitter();
    }
    PellComponent.prototype.ngOnInit = function () {
    };
    PellComponent.prototype.onInput = function (event) {
        this.onChange.emit(event.target.innerHTML);
    };
PellComponent.ɵfac = function PellComponent_Factory(t) { return new (t || PellComponent)(); };
PellComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: PellComponent, selectors: [["pell-editor"]], outputs: { onChange: "onChange" }, decls: 5, vars: 5, consts: [["pell", ""], [3, "class", "title", "innerHTML", "click", 4, "ngFor", "ngForOf"], ["contenteditable", "true", 3, "input"], [3, "title", "innerHTML", "click"]], template: function PellComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", null, 0);
        ɵngcc0.ɵɵelementStart(2, "div");
        ɵngcc0.ɵɵtemplate(3, PellComponent_button_3_Template, 1, 4, "button", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 2);
        ɵngcc0.ɵɵlistener("input", function PellComponent_Template_div_input_4_listener($event) { return ctx.onInput($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵclassMap(ctx.actionBarClass);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.actions);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassMap(ctx.contentClass);
    } }, directives: [ɵngcc1.NgForOf], styles: [".pell[_ngcontent-%COMP%]{border-radius:5px;-webkit-box-shadow:0 2px 3px hsla(0,0%,4%,.1),0 0 0 1px hsla(0,0%,4%,.1);box-shadow:0 2px 3px hsla(0,0%,4%,.1),0 0 0 1px hsla(0,0%,4%,.1)}.pell[_ngcontent-%COMP%], .pell-content[_ngcontent-%COMP%]{-webkit-box-sizing:border-box;box-sizing:border-box;width:100%}.pell-content[_ngcontent-%COMP%]{height:300px;outline:0;overflow-y:auto;padding:10px}.pell-actionbar[_ngcontent-%COMP%]{background-color:#fff;border-bottom:1px solid hsla(0,0%,4%,.1);border-top-left-radius:5px;border-top-right-radius:5px;width:100%}.pell-button[_ngcontent-%COMP%]{background-color:transparent;border:none;cursor:pointer;height:30px;outline:0;width:30px}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PellComponent, [{
        type: core.Component,
        args: [{
                selector: 'pell-editor',
                template: "<div #pell>\n  <div [class]=\"actionBarClass\">\n    <button *ngFor=\"let action of actions\" [class]=\"actionButtonClass\" [title]=\"action.title\"\n    (click)=\"action.result()\" [innerHTML]=\"action.icon\">\n    </button>\n  </div>\n  <div contenteditable=\"true\" [class]=\"contentClass\" (input)=\"onInput($event)\"></div>\n</div>\n",
                styles: [".pell{border-radius:5px;-webkit-box-shadow:0 2px 3px hsla(0,0%,4%,.1),0 0 0 1px hsla(0,0%,4%,.1);box-shadow:0 2px 3px hsla(0,0%,4%,.1),0 0 0 1px hsla(0,0%,4%,.1)}.pell,.pell-content{-webkit-box-sizing:border-box;box-sizing:border-box;width:100%}.pell-content{height:300px;outline:0;overflow-y:auto;padding:10px}.pell-actionbar{background-color:#fff;border-bottom:1px solid hsla(0,0%,4%,.1);border-top-left-radius:5px;border-top-right-radius:5px;width:100%}.pell-button{background-color:transparent;border:none;cursor:pointer;height:30px;outline:0;width:30px}"]
            }]
    }], function () { return []; }, { onChange: [{
            type: core.Output
        }] }); })();
    return PellComponent;
}());
PellComponent.ctorParameters = function () { return []; };
PellComponent.propDecorators = {
    "onChange": [{ type: core.Output },],
};
var PellModule = /** @class */ (function () {
    function PellModule() {
    }
PellModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: PellModule });
PellModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function PellModule_Factory(t) { return new (t || PellModule)(); }, imports: [[
            common.CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(PellModule, { declarations: [PellComponent], imports: [ɵngcc1.CommonModule], exports: [PellComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PellModule, [{
        type: core.NgModule,
        args: [{
                imports: [
                    common.CommonModule
                ],
                declarations: [PellComponent],
                exports: [PellComponent]
            }]
    }], function () { return []; }, null); })();
    return PellModule;
}());
PellModule.ctorParameters = function () { return []; };

exports.PellModule = PellModule;
exports.ɵa = PellComponent;

Object.defineProperty(exports, '__esModule', { value: true });

})));
