// angular2-tinymce/esm5/lib/angular2-tinymce-lib.default.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
var TinymceDefaultOptions = /** @class */ (function () {
    function TinymceDefaultOptions() {
        this.plugins = [
            'link',
            'paste',
            'table',
            'advlist',
            'autoresize',
            'lists',
            'code'
        ];
        this.skin_url = '/assets/tinymce/skins/lightgray';
        this.baseURL = '/assets/tinymce';
        this.auto_focus = true;
    }
TinymceDefaultOptions.ɵfac = function TinymceDefaultOptions_Factory(t) { return new (t || TinymceDefaultOptions)(); };
TinymceDefaultOptions.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: TinymceDefaultOptions, factory: function (t) { return TinymceDefaultOptions.ɵfac(t); } });

    return TinymceDefaultOptions;
}());
export { TinymceDefaultOptions };
if (false) {
    /** @type {?} */
    TinymceDefaultOptions.prototype.plugins;
    /** @type {?} */
    TinymceDefaultOptions.prototype.skin_url;
    /** @type {?} */
    TinymceDefaultOptions.prototype.baseURL;
    /** @type {?} */
    TinymceDefaultOptions.prototype.auto_focus;
}
