// angular-highcharts-chart/bundles/angular-highcharts-chart.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('angular-highcharts-chart', ['@angular/core','exports', '@angular/core'], factory) :
    (factory(global.ng.core,(global['angular-highcharts-chart'] = {}),global.ng.core));
}(this, (function (ɵngcc0,exports,core) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var AngularHighchartsChartComponent = (function () {
        function AngularHighchartsChartComponent(el) {
            this.el = el;
            this.updateValue = false;
            this.updateChange = new core.EventEmitter(true);
            this.updateOrCreateChart = function () {
                if (this.chart && this.chart.update) {
                    this.chart.update(this.optionsValue, true, this.oneToOne || false);
                }
                else {
                    this.chart = this.Highcharts[this.constructorType || 'chart'](this.el.nativeElement, this.optionsValue, this.callbackFunction || null);
                    this.optionsValue.series = this.chart.userOptions.series;
                }
            };
        }
        Object.defineProperty(AngularHighchartsChartComponent.prototype, "options", {
            set: /**
             * @param {?} val
             * @return {?}
             */ function (val) {
                this.optionsValue = val;
                this.updateOrCreateChart();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AngularHighchartsChartComponent.prototype, "update", {
            set: /**
             * @param {?} val
             * @return {?}
             */ function (val) {
                if (val) {
                    this.updateOrCreateChart();
                    this.updateChange.emit(false); // clear the flag after update
                }
            },
            enumerable: true,
            configurable: true
        });
        /** @nocollapse */
        AngularHighchartsChartComponent.ctorParameters = function () {
            return [
                { type: core.ElementRef }
            ];
        };
        AngularHighchartsChartComponent.propDecorators = {
            Highcharts: [{ type: core.Input }],
            constructorType: [{ type: core.Input }],
            callbackFunction: [{ type: core.Input }],
            options: [{ type: core.Input }],
            updateChange: [{ type: core.Output }],
            update: [{ type: core.Input }],
            oneToOne: [{ type: core.Input }]
        };
AngularHighchartsChartComponent.ɵfac = function AngularHighchartsChartComponent_Factory(t) { return new (t || AngularHighchartsChartComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
AngularHighchartsChartComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AngularHighchartsChartComponent, selectors: [["lib-highcharts-angular"]], inputs: { options: "options", update: "update", Highcharts: "Highcharts", constructorType: "constructorType", callbackFunction: "callbackFunction", oneToOne: "oneToOne" }, outputs: { updateChange: "updateChange" }, decls: 0, vars: 0, template: function AngularHighchartsChartComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularHighchartsChartComponent, [{
        type: core.Component,
        args: [{
                selector: 'lib-highcharts-angular',
                template: ''
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { updateChange: [{
            type: core.Output
        }], options: [{
            type: core.Input
        }], update: [{
            type: core.Input
        }], Highcharts: [{
            type: core.Input
        }], constructorType: [{
            type: core.Input
        }], callbackFunction: [{
            type: core.Input
        }], oneToOne: [{
            type: core.Input
        }] }); })();
        return AngularHighchartsChartComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var AngularHighchartsChartModule = (function () {
        function AngularHighchartsChartModule() {
        }
AngularHighchartsChartModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularHighchartsChartModule });
AngularHighchartsChartModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularHighchartsChartModule_Factory(t) { return new (t || AngularHighchartsChartModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularHighchartsChartModule, { declarations: [AngularHighchartsChartComponent], exports: [AngularHighchartsChartComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularHighchartsChartModule, [{
        type: core.NgModule,
        args: [{
                imports: [],
                declarations: [AngularHighchartsChartComponent],
                exports: [AngularHighchartsChartComponent]
            }]
    }], function () { return []; }, null); })();
        return AngularHighchartsChartModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    // export * from './lib/angular-highcharts-chart.component';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.AngularHighchartsChartModule = AngularHighchartsChartModule;
    exports.ɵa = AngularHighchartsChartComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
